XXX-asyncwrap-api.md

```diff
@@ -0,0 +1,628 @@
+request batches calls under the hood this would still allow for each request
+to trigger the callbacks individually so that the callback associated with
+each bach request execute inside the correct parent.
+
+This cannot be enforced in any way. It is up to the implementer to make sure
+all of the callbacks are placed and called at the correct time.
+
+
+#### `async_wrap.createEvent(handle[, parent_id])`
+
+* `handle` {Object}
+* `parent_id` {Number} **Default:** `async_wrap.currentId()`
+* Return: {AsyncEvent}
+
+Returns an instance of `AsyncEvent` to be used to trigger `AsyncHook` callback.
+
+`handle` is the object instance that's passed to `init()`.
+
+`parent_id` is the value passed as `parentId` to `init()`. If omitted then the
+current async id is used.
+
+An instance of `AsyncEvent` is returned that is to be used to trigger the
+`AsyncHook` callbacks.
+
+
+#### `event.getId()`
+
+* Return: {Number}
+
+Returns the unique id assigned to the instance when it was created.
+
+Would be easier to just slap on the number, but right now I'm not 100% sure
+that the implementation won't change in the future. So sticking with this.
+
+
+#### `event.emitInit()`
+
+* Return: {Undefined}
+
+Trigger listening `init()` callbacks that this event has been created.  Will
+throw if `emitInit()` is called twice.
+
+
+#### `event.emitBefore()`
+
+* Return: {Undefined}
+
+Trigger listening `before()` callbacks that the `handle` is about to enter it's
+call stack.
+
+
+#### `event.emitAfter()`
+
+* Return: {AsyncEvent}
+
+Trigger listening `after()` callbacks that the `handle` has left it's call
+stack.
+
+The `id` call stack should always properly unwind. No two `emitAfter()` calls
+should be crossed. For example:
+
+```
+init # Foo
+init # Bar
+...
+before # Foo
+before # Bar
+after # Foo   <- Should be called after Bar
+after # Bar
+```
+
+There is no known scenario where this is acceptable. If one is found then this
+restriction may be lifted.
+
+Return value is the calling `AsyncEvent` instance. This is done to make the
+following scenario more simple:
+
+```
+event.emitAfter().emitDestroy();
+```
+
+Because it won't be uncommon that the `event` will be destroyed immediately
+following the `after()` callback has been called.
+
+
+#### `event.emitDestroy()`
+
+* Return: {Undefined}
+
+Trigger listening `destroy()` callbacks that the `handle`'s resources are no
+longer being used, and in cases where the `handle` is attached to native
+resources the underlying class may have been destructed.
+
+
+## API Exceptions
+
+### net Client connection Event
+
+Technically the `before()`/`after()` events of the `'connection'` event for
+`net.Server` would place the server as the active id. Problem is that this is
+not intuitive in how the asynchronous chain would propagate. So instead make
+the client the active id for the duration of the `'connection'` callback.
+
+
+### Reused Resources
+
+Resources like `HTTPParser` are reused throughout the lifetime of the
```

Comment:
The public docs will have one.

The big ones that you would commonly run into would be `HTTPParser` and `TimerWrap`.